libraries/libstratosphere/source/ncm/ncm_content_meta_database_impl.cpp

```diff
@@ -0,0 +1,440 @@
+        return ResultSuccess();
+    }
+
+    Result ContentMetaDatabaseImpl::GetRequiredSystemVersion(sf::Out<u32> out_version, const ContentMetaKey &key) {
+        R_TRY(this->EnsureEnabled());
+
+        /* Only applications and patches have a required system version. */
+        R_UNLESS(key.type == ContentMetaType::Application || key.type == ContentMetaType::Patch, ncm::ResultInvalidContentMetaKey());
+
+        /* Obtain the content meta for the key. */
+        const void *meta;
+        size_t meta_size;
+        R_TRY(this->GetContentMetaPointer(&meta, &meta_size, key));
+
+        /* Create a reader. */
+        ContentMetaReader reader(meta, meta_size);
+
+        /* Obtain the required system version. */
+        switch (key.type) {
+            case ContentMetaType::Application:
+                out_version.SetValue(reader.GetExtendedHeader<ApplicationMetaExtendedHeader>()->required_system_version);
+                break;
+            case ContentMetaType::Patch:
+                out_version.SetValue(reader.GetExtendedHeader<PatchMetaExtendedHeader>()->required_system_version);
+                break;
+            AMS_UNREACHABLE_DEFAULT_CASE();
+        }
+
+        return ResultSuccess();
+    }
+
+    Result ContentMetaDatabaseImpl::GetPatchId(sf::Out<PatchId> out_patch_id, const ContentMetaKey &key) {
+        R_TRY(this->EnsureEnabled());
+
+        /* Only applications can have patches. */
+        R_UNLESS(key.type == ContentMetaType::Application, ncm::ResultInvalidContentMetaKey());
+
+        /* Obtain the content meta for the key. */
+        const void *meta;
+        size_t meta_size;
+        R_TRY(this->GetContentMetaPointer(&meta, &meta_size, key));
+
+        /* Create a reader. */
+        ContentMetaReader reader(meta, meta_size);
+
+        /* Obtain the patch id. */
+        out_patch_id.SetValue(reader.GetExtendedHeader<ApplicationMetaExtendedHeader>()->patch_id);
+        return ResultSuccess();
+    }
+
+    Result ContentMetaDatabaseImpl::DisableForcibly() {
+        this->disabled = true;
+        return ResultSuccess();
+    }
+
+    Result ContentMetaDatabaseImpl::LookupOrphanContent(const sf::OutArray<bool> &out_orphaned, const sf::InArray<ContentId> &content_ids) {
+        R_TRY(this->EnsureEnabled());
+        R_UNLESS(out_orphaned.GetSize() >= content_ids.GetSize(), ncm::ResultBufferInsufficient());
+
+        /* Default to orphaned for all content ids. */
+        for (size_t i = 0; i < out_orphaned.GetSize(); i++) {
+            out_orphaned[i] = true;
+        }
+
+        auto IsOrphanedContent = [](const sf::InArray<ContentId> &list, const ncm::ContentId &id) ALWAYS_INLINE_LAMBDA {
+            /* Check if any input content ids match our found content id. */
+            for (size_t i = 0; i < list.GetSize(); i++) {
+                if (list[i] == id) {
+                    return std::make_optional(i);
+                }
+            }
+            return std::optional<size_t>(std::nullopt);
```

Comment:
This syntax is required to workaround gcc return type deduction bugs. The above won't compile without it

Reply:
Ah, I see, that's unfortunate.

Reply:
@leoetlino Note that gcc 10 fixes a gcc bug for ALWAYS_INLINE attribute in conjunction with trailing return types. Once we're on GCC 10 this can be done (and we can just do -> std::optional<size_t> in the lambda def).

I'll add a TODO C++20 note, as I intend to address everything in the codebase tagged TODO C++20 when we upgrade to GCC 10 in a few months.